command/var_list_test.go

```diff
@@ -0,0 +1,511 @@
+			if tC.expectStdErrPrefix != "" {
+				require.True(t, strings.HasPrefix(errOut, tC.expectStdErrPrefix),
+					"Expected stderr to start with %q; got %s",
+					tC.expectStdErrPrefix, errOut)
+				// Test that stderr ends with a linefeed since we trim them for
+				// convenience in the equality tests.
+				require.True(t, strings.HasSuffix(errOut, "\n"),
+					"stderr does not end with a linefeed")
+			}
+		})
+	}
+}
+
+// TestVarListCommand_Online contains all of the tests that use a testagent.
+// They reuse the same testagent so that they can run in parallel and minimize
+// test startup time costs.
+func TestVarListCommand_Online(t *testing.T) {
+	ci.Parallel(t)
+
+	// Create a server
+	srv, client, url := testServer(t, true, nil)
+	defer srv.Shutdown()
+
+	ui := cli.NewMockUi()
+	cmd := &VarListCommand{Meta: Meta{Ui: ui}}
+
+	nsList := []string{api.DefaultNamespace, "ns1"}
+	pathList := []string{"a/b/c", "a/b/c/d", "z/y", "z/y/x"}
+	toJSON := func(in interface{}) string {
+		b, err := json.MarshalIndent(in, "", "    ")
+		if err != nil {
+			return ""
+		}
+		return strings.TrimSpace(string(b))
+	}
+	variables := setupTestVariables(client, nsList, pathList)
+
+	testTmpl := `{{ range $i, $e := . }}{{if ne $i 0}}{{print "•"}}{{end}}{{printf "%v\t%v" .Namespace .Path}}{{end}}`
+
+	pathsEqual := func(t *testing.T, expect any) testVarListJSONTestExpectFn {
+		out := func(t *testing.T, check any) {
+
+			expect := expect
+			exp, ok := expect.(NSPather)
+			require.True(t, ok, "expect is not an NSPather, got %T", expect)
+			in, ok := check.(NSPather)
+			require.True(t, ok, "check is not an NSPather, got %T", check)
+			require.ElementsMatch(t, exp.NSPaths(), in.NSPaths())
+		}
+		return out
+	}
+
+	hasLength := func(t *testing.T, length int) testVarListJSONTestExpectFn {
+		out := func(t *testing.T, check any) {
+
+			length := length
+			in, ok := check.(NSPather)
+			require.True(t, ok, "check is not an NSPather, got %T", check)
+			inLen := in.NSPaths().Len()
+			require.Equal(t, length, inLen,
+				"expected length of %v, got %v. \nvalues: %v",
+				length, inLen, in.NSPaths())
+		}
+		return out
+	}
+
+	testCases := []testVarListTestCase{
+		{
+			name:         "plaintext/not found",
+			args:         []string{"does/not/exist"},
+			expectStdOut: msgSecureVariableNotFound,
+		},
+		{
+			name: "plaintext/single variable",
+			args: []string{"a/b/c/d"},
+			expectStdOut: formatList([]string{
+				"Namespace|Path|Last Updated",
+				fmt.Sprintf(
+					"default|a/b/c/d|%s",
+					time.Unix(0, variables.HavingPrefix("a/b/c/d")[0].ModifyTime),
+				),
+			},
+			),
+		},
+		{
+			name:         "plaintext/quiet",
+			args:         []string{"-q"},
+			expectStdOut: strings.Join(variables.HavingNamespace(api.DefaultNamespace).Strings(), "\n"),
+		},
+		{
+			name:         "plaintext/quiet/prefix",
+			args:         []string{"-q", "a/b/c"},
+			expectStdOut: strings.Join(variables.HavingNSPrefix(api.DefaultNamespace, "a/b/c").Strings(), "\n"),
```

Comment:
I don't think this is a blocker but maybe more for future reference: I feel like having logic spread across some new test-only structs and the return value of the variable setup code, etc. muddies the tests. The setup code isn't concurrent and all the paths are known in advance. What stops us from having the raw expected output here, maybe with test helper to strip off the timestamps? Then the test case reads fairly straightforwardly without having to refer to code at both the top and bottom of the page to know what we're doing.